REST-API-Basics/service/src/main/java/com/epam/esm/service/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,212 @@
+package com.epam.esm.service.service.impl;
+
+import com.epam.esm.model.constant.SortParamsContext;
+import com.epam.esm.model.dao.GiftCertificateDao;
+import com.epam.esm.model.dao.TagDao;
+import com.epam.esm.model.entity.GiftCertificate;
+import com.epam.esm.model.entity.Tag;
+import com.epam.esm.service.dto.GiftCertificateDto;
+import com.epam.esm.service.exception.ServiceException;
+import com.epam.esm.service.service.GiftCertificateService;
+import com.epam.esm.service.validator.Validator;
+import com.epam.esm.service.validator.impl.GiftCertificateValidatorImpl;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.math.BigDecimal;
+import java.util.*;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final GiftCertificateDao giftCertificateDao;
+    private final TagDao tagDao;
+    private final Validator<GiftCertificate> giftCertificateValidator;
+    private final Validator<Tag> tagValidator;
+
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, TagDao tagDao,
+                                      Validator<GiftCertificate> giftCertificateValidator,
+                                      Validator<Tag> tagValidator) {
+        this.giftCertificateDao = giftCertificateDao;
+        this.tagDao = tagDao;
+        this.giftCertificateValidator = giftCertificateValidator;
+        this.tagValidator = tagValidator;
+
+    }
+
+    @Override
+    public List<GiftCertificate> getAll() {
+        return giftCertificateDao.getAll();
+    }
+
+    @Override
+    @Transactional
+    public void create(GiftCertificateDto giftCertificateDto) throws ServiceException {
+        GiftCertificate giftCertificate = giftCertificateDto.getGiftCertificate();
+        validateGiftCertificate(giftCertificate);
+        validateTags(giftCertificateDto.getCertificateTags());
+        String giftCertificateName = giftCertificate.getName();
+        boolean isCertificateExist = giftCertificateDao.findByName(giftCertificateName).isPresent();
+        if (isCertificateExist) {
+            throw new ServiceException("message.giftCertificateExists");
+        }
+        giftCertificateDao.create(giftCertificate);
+        Long giftCertificateId = giftCertificateDao.findByName(giftCertificateName)
```

Comment:
if id = -1 means that certificate wasn't created successfully after all passed validation checks, better throw an exception instead of continuing method execution